proposals/0000-swift-settings.md

```diff
@@ -0,0 +1,446 @@
+  contains mostly UI elements but that wishes to implement helper abstractions
+  that interact with a database on a background task. This helper code is
+  naturally expressed as having a default isolation of nonisolated implying that
+  one must either split the database helpers into a separate module or must mark
+  most declarations in the database code with nonisolated by hand.
+
+## Proposed solution
+
+We propose introducing a new macro called `#SwiftSettings` that takes a list of
+structs of type `SwiftSetting`. `SwiftSetting` will possess a static method for
+each command line argument that it supports. Any parameters that would be passed
+on the command line are passed as parameters to the static method. The compiler
+upon parsing a `#SwiftSettings` macro updates its internal state to set the
+appropriate settings before compiling the rest of the file. An example of such a
+`#SwiftSettings` invocation is the following:
+
+```swift
+#SwiftSettings(
+    .defaultIsolation(MainActor.self)
+)
+```
+
+By specifying compiler options in such a manner, we are able to avoid the need
+for module splitting if a user wishes to change the default isolation only on a
+specific file.
+
+As an additional benefit since strongly typed methods are being used instead of
+providing a single method that takes a string to specify the command line
+parameter code completion can make the compiler options discoverable to the user
+in the IDE.
+
+NOTE: Initially we will only support default isolation. In the future, this can
+be loosened as appropriate (see future directions).
+
+## Detailed design
+
+We will introduce into the standard library a new declaration macro called
+`#SwiftSettings` and a new struct called `SwiftSetting` that `#SwiftSettings`
+takes as a variadic parameter:
+
+```swift
+@freestanding(declaration)
+public macro SwiftSettings(_ settings: SwiftSetting...) = Builtin.SwiftSettingsMacro
+
+public struct SwiftSetting {
+  public init() { fatalError("Cannot construct an instance of SwiftSetting") }
+}
+```
+
+`#SwiftSettings` will expand to an empty string and is used only for its syntax
+and documentation in the IDE.
+
+`SwiftSetting` is a resilient struct that cannot be constructed without a
+runtime error and is only used for the purposes of specifying command line
+arguments to a `#SwiftSettings` macro. Support for individual command line
+options is added to `SwiftSetting` by defining a static method on `SwiftSetting`
+in an extension. By using static methods and extensions in this manner, we are
+able to make `SwiftSetting` extensible without requiring the standard library to
+be updated whenever a new options is added since the extension can be defined in
+other modules. As an example, we add support for `defaultIsolation` via an
+extension in the Concurrency module as follows:
+
+```swift
+extension SwiftSetting {
+  public static func defaultIsolation(_ isolation: Actor.Type?) -> SwiftSetting { SwiftSetting() }
```

Comment:
This should match the package manifest API in SE-0466, which uses `MainActor.Type?`